Lib/test/pickletester.py

```diff
@@ -2994,6 +2995,70 @@ def __reduce__(self):
 class BBB(object):
     pass
 
+
+class AbstractHookTests(unittest.TestCase):
+    def test_pickler_hook(self):
+        # test the CPickler ability to register custom, user-defined reduction
+        # callbacks to pickle functions and classes.
+
+        def custom_reduction_callback(pickler, obj):
+            obj_name = obj.__name__
+
+            if obj_name == 'f':
+                # asking the pickler to save f as 5
+                return int, (5, )
+
+            if obj_name == 'MyClass':
+                return str, ('some str',)
+
+            elif obj_name == 'log':
+                # letting the pickler falling back to buitlin save_global to
+                # pickle functions named 'log' by attribute.
+                return NotImplemented
+
+            elif obj_name == 'g':
+                # in this case, the callback returns an invalid result (not a
+                # 2-5 tuple), the pickler should raise a proper error.
+                return False
+            elif obj_name == 'h':
+                # Simulate a case when the reducer fails. The error should be
+                # propagated to the original ``dump`` call.
+                raise ValueError('The reducer just failed')
+            return NotImplemented
+
+        def f():
+            pass
+
+        def g():
+            pass
+
+        def h():
+            pass
+
+        class MyClass:
+            pass
+
+        for proto in range(0, pickle.HIGHEST_PROTOCOL + 1):
+            with self.subTest(proto=proto):
+                bio = io.BytesIO()
+                p = self.pickler_class(bio, proto)
+                p.reducer_override = custom_reduction_callback
```

Comment:
Would be nice if once could also define the hook as a method:
```python
class MyPickle(pickle.Pickler):
    def reducer_override(self, obj):
        ...
```

Reply:
I changed it, but not "ALSO" as a method, but "UNIQUELY as a method". To me, it should be enough.